alacritty/src/display/mod.rs

```diff
@@ -1534,6 +1560,63 @@ impl<T> DerefMut for Replaceable<T> {
     }
 }
 
+/// The frame timer state.
+pub struct FrameTimerInfo {
+    /// The optimal delay for the renderer.
+    swap_timeout: Duration,
+
+    /// The previous error that we've used to offset the `swap_delay`.
+    previous_error: i64,
+
+    /// The timestamp of the previous frame.
+    frame_timestamp: Instant,
+
+    /// The refresh rate we've used to compute the `swap_delay`.
+    refresh_interval: Duration,
+}
+
+impl FrameTimerInfo {
+    pub fn new() -> Self {
+        Self {
+            swap_timeout: Duration::ZERO,
+            previous_error: 0,
+            frame_timestamp: Instant::now(),
+            refresh_interval: Duration::ZERO,
+        }
+    }
+
+    /// Compute the delay that we should use to achieve the target frame
+    /// rate.
+    pub fn compute_timeout(&mut self, refresh_interval: Duration) -> Duration {
+        // Handle refresh rate change.
+        if self.refresh_interval != refresh_interval {
+            self.refresh_interval = refresh_interval;
+            self.frame_timestamp = Instant::now();
+            self.swap_timeout = refresh_interval;
+            return refresh_interval;
+        }
+
+        // Compute the error.
+        let error = self.refresh_interval.as_micros() as i64
+            - self.frame_timestamp.elapsed().as_micros() as i64;
+
+        // Don't update error and swap timeout if we've waited for too long.
+        if error < self.refresh_interval.as_micros() as i64 {
+            // Average previously used error with the current one to reduce momentum from the
+            // error sign changes.
+            self.previous_error = (self.previous_error + error) / 2;
+
+            let new_swap_timeout = self.swap_timeout.as_micros() as i64 + self.previous_error;
```

Comment:
Hm I'm a little concerned with this. Previously since we adjusted for the error **exactly** we'd obviously always draw in `swap_timeout` intervals. But now that we adjust for this error, we have no reference point for our "default" clock cycle anymore. So we'd drift off course since the error correction with the previous frame is not precise.

Instead of taking the `swap_timeout` I feel like we should be calculating the time until the next clock cycle (based on start time) and use that instead?

Reply:
>point for our "default" clock cycle anymore. So we'd drift off course since the error correction with the previous frame is not precise.

We use `refresh_interval` for that? So we have a reference, we just try to converge to it slower.

>Instead of taking the swap_timeout I feel like we should be calculating the time until the next clock cycle (based on start time) and use that instead?

Could you elaborate on that? You mean that? For now we use the refresh interval to cap us at refresh rate, do you want some arbitrary logic like get `Instant::now()` at startup, and then draw only on  `start + X * modulo`? I'm not sure it'll be any better to be honest...

Reply:
> We use refresh_interval for that? So we have a reference, we just try to converge to it slower.

You have an interval that converges using two historical datapoints. But let's assume this averaging has introduced a 1 microsecond drift in one direction due to rounding. You're never accounting for that drift in the future since all you're doing is just taking the last two errors and averaging them. As a result a drift of 1 microsecond is introduced, shifting your clock permanently.

> Could you elaborate on that? You mean that? For now we use the refresh interval to cap us at refresh rate, do you want some arbitrary logic like get Instant::now() at startup, and then draw only on start + X * modulo? I'm not sure it'll be any better to be honest...

I'm not sure about the exact implementation or your suggestion, but basically yeah. We're already using `elapsed()` to calculate the error, what I'd suggest basically is just doing the same but keeping our clock consistent with the start of the clock rather than consistent with the last frame. That should improve frame consistency.